File: app/src/main/java/com/moggot/findmycarlocation/di/module/NetworkModule.java
package com.moggot.findmycarlocation.di.module;

import androidx.annotation.NonNull;

import com.moggot.findmycarlocation.data.api.LocationApi;
import com.moggot.findmycarlocation.data.repository.local.SettingsPreferences;
import com.moggot.findmycarlocation.data.repository.network.NetworkRepo;
import com.moggot.findmycarlocation.di.scope.MainScope;
import com.moggot.findmycarlocation.retry.NetworkRetryManager;
import com.moggot.findmycarlocation.retry.RetryManager;

import java.util.concurrent.Executor;
import java.util.concurrent.Executors;

import dagger.Module;
import dagger.Provides;
import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;
import timber.log.Timber;

@Module
public class NetworkModule {

    @Provides
    @MainScope
    HttpLoggingInterceptor provideHttpLoggingInterceptor() {
        HttpLoggingInterceptor httpLoggingInterceptor =
                new HttpLoggingInterceptor(message -> Timber.d(message));
        httpLoggingInterceptor.setLevel(HttpLoggingInterceptor.Level.BASIC);
        return httpLoggingInterceptor;
    }

    @Provides
    @MainScope
    OkHttpClient provideOkHttpClient(HttpLoggingInterceptor loggingInterceptor) {
        return new OkHttpClient.Builder()
                .addInterceptor(loggingInterceptor)
                .build();
    }

    @Provides
    @MainScope
    Retrofit.Builder provideRetrofitBuilder(OkHttpClient client) {
        return new Retrofit.Builder()
                .client(client)
                .addConverterFactory(GsonConverterFactory.create())
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create());
    }

    @Provides
    @MainScope
    LocationApi provideLocationApi(Retrofit.Builder builder) {
        return builder
                .baseUrl(LocationApi.BASE_LOCATION_URL)
                .build()
                .create(LocationApi.class);
    }

    @Provides
    @MainScope
    NetworkRepo provideNetworkService(@NonNull LocationApi locationApi, @NonNull SettingsPreferences preferences) {
        return new NetworkRepo(locationApi, preferences);
    }

    @Provides
    @MainScope
    Executor provideExecutor() {
        return Executors.newFixedThreadPool(2);
    }

    @Provides
    @MainScope
    RetryManager provideRetryManager() {
        return new NetworkRetryManager();
    }
}

File: app/src/main/java/com/moggot/findmycarlocation/di/module/NetworkModule.kt
package com.moggot.findmycarlocation.di.module

import com.moggot.findmycarlocation.App
import com.moggot.findmycarlocation.data.api.LocationApi
import dagger.Module
import dagger.Provides
import dagger.hilt.InstallIn
import dagger.hilt.components.SingletonComponent
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.converter.moshi.MoshiConverterFactory
import timber.log.Timber
import javax.inject.Singleton

@Module
@InstallIn(SingletonComponent::class)
class NetworkModule {

    @Provides
    @Singleton
    fun provideHttpLoggingInterceptor(): HttpLoggingInterceptor {
        val httpLoggingInterceptor =
            HttpLoggingInterceptor { message -> Timber.tag(App.TAG).d("HTTP REQUEST: $message") }
        httpLoggingInterceptor.setLevel(HttpLoggingInterceptor.Level.BASIC)
        return httpLoggingInterceptor
    }

    @Provides
    @Singleton
    fun provideOkHttpClient(loggingInterceptor: HttpLoggingInterceptor): OkHttpClient {
        return OkHttpClient.Builder()
            .addInterceptor(loggingInterceptor)
            .build()
    }

    @Provides
    @Singleton
    fun provideRetrofitBuilder(client: OkHttpClient): Retrofit.Builder {
        return Retrofit.Builder()
            .client(client)
            .addConverterFactory(MoshiConverterFactory.create())
    }

    @Provides
    @Singleton
    fun provideLocationApi(builder: Retrofit.Builder): Retrofit {
        return builder
            .baseUrl(LocationApi.BASE_LOCATION_URL)
            .build()
    }
}